org-roam-doctor.el

```diff
@@ -0,0 +1,199 @@
+               ("r" . ("Replace link" . org-roam-doctor--replace-link))
+               ("R" . ("Replace link (keep label)" . org-roam-doctor--replace-link-keep-label))))))
+
+(defun org-roam-doctor-broken-links (ast)
+  "Checker for detecting broken links.
+AST is the org-element parse tree."
+  (org-element-map ast 'link
+    (lambda (l)
+      (when (equal "file" (org-element-property :type l))
+        (let ((file (org-element-property :path l)))
+          (and (not (file-remote-p file))
+               (not (file-exists-p file))
+               (list (org-element-property :begin l)
+                     (format (if (org-element-lineage l '(link))
+                                 "Link to non-existent image file \"%s\"\
+ in link description"
+                               "Link to non-existent local file \"%s\"")
+                             file))))))))
+
+(defun org-roam-doctor--check (buffer checkers)
+  "Check BUFFER for errors.
+CHECKERS is the list of checkers used."
+  (with-current-buffer buffer
+    (save-excursion
+      (goto-char (point-min))
+      (let* ((ast (org-element-parse-buffer))
+             (errors (sort (cl-mapcan
+                            (lambda (c)
+                              (mapcar
+                               (lambda (report)
+                                 (list (set-marker (make-marker) (car report))
+                                       (nth 1 report) c))
+                               (save-excursion
+                                 (funcall
+                                  (org-roam-doctor-checker-name c)
+                                  ast))))
+                            checkers)
+                           #'car-less-than-car)))
+        (dolist (e errors)
+          (pcase-let ((`(,m ,msg ,checker) e))
+            (switch-to-buffer buffer)
+            (goto-char m)
+            (org-reveal)
+            (org-roam-doctor--resolve msg checker)
+            (set-marker m nil)))
+        errors))))
+
+;;; Actions
+(defun org-roam-doctor--recursive-edit ()
+  "Launch into a recursive edit."
+  (message "When you're done editing press C-M-c to continue.")
+  (recursive-edit))
+
+(defun org-roam-doctor--replace-link ()
+  "Replace the current link with a new link."
+  (unless (org-in-regexp org-link-bracket-re 1)
+    (user-error "No link at point"))
+  (save-excursion
+    (delete-region (match-beginning 0) (match-end 0))
+    (org-roam-insert)))
+
+(defun org-roam-doctor--replace-link-keep-label ()
+  "Replace the current link with a new link, keeping the current link's label."
+  (unless (org-in-regexp org-link-bracket-re 1)
+    (user-error "No link at point"))
+  (save-excursion
+    (let ((label (if (match-end 2)
+                     (match-string-no-properties 2)
+                   (org-link-unescape (match-string-no-properties 1)))))
+      (delete-region (match-beginning 0) (match-end 0))
+      (org-roam-insert nil nil label))))
```

Comment:
If we want to keep the label, chances are we might be interested in pre-selecting any candidate that starts with that label.  For instance, if I have a broken link with the label `Metaphor`, fixing it should pre-select `Metaphor` in my list.  This could be done by adding a `PRE-SELECT` argument to `org-roam-insert` which, when set, would attempt to place the matched candidate at the top of the completions.  Sorting also has the benefit of being completion-engine-agnostic.

Reply:
can this be done with regular `completing-read`? Not familiar with how to do this.